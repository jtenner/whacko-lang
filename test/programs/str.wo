export fn _start(): void {
  let a = "Hello world";
}