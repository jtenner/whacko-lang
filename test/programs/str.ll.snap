; ModuleID = 'whacko'
source_filename = "whacko"
target triple = "wasm32-wasi"

define void @_start() {
entry:
  %"alloca~0.i" = alloca ptr, align 8
  %"alloca~1.i" = alloca i32, align 4
  %"alloca~2.i" = alloca i32, align 4
  %"alloca~0" = alloca ptr, align 8
  %"36~inst" = tail call ptr @malloc(i32 ptrtoint (ptr getelementptr (ptr, ptr null, i32 1) to i32))
  store i32 1, ptr %"36~inst", align 4
  %"sizeOfGEP~53" = getelementptr ptr, ptr %"36~inst", i32 1
  store i32 ptrtoint (ptr getelementptr (ptr, ptr null, i32 1) to i32), ptr %"sizeOfGEP~53", align 4
  call void @llvm.lifetime.start.p0(i64 8, ptr %"alloca~0.i")
  call void @llvm.lifetime.start.p0(i64 4, ptr %"alloca~1.i")
  call void @llvm.lifetime.start.p0(i64 4, ptr %"alloca~2.i")
  store ptr %"36~inst", ptr %"alloca~0.i", align 8
  store i32 ptrtoint ([12 x i8] c"Hello world\00" to i32), ptr %"alloca~1.i", align 4
  store i32 11, ptr %"alloca~2.i", align 4
  %"gep~54.i" = getelementptr { i32, i32, i32, i32 }, ptr %"alloca~0.i", i32 2
  %"45~inst.i" = load i32, ptr %"gep~54.i", align 4
  %"46~inst.i" = load i32, ptr %"alloca~1.i", align 4
  %"gep~55.i" = getelementptr { i32, i32, i32, i32 }, ptr %"alloca~0.i", i32 2
  store i32 %"46~inst.i", ptr %"gep~55.i", align 4
  %"gep~56.i" = getelementptr { i32, i32, i32, i32 }, ptr %"alloca~0.i", i32 3
  %"48~inst.i" = load i32, ptr %"gep~56.i", align 4
  %"49~inst.i" = load i32, ptr %"alloca~2.i", align 4
  %"gep~57.i" = getelementptr { i32, i32, i32, i32 }, ptr %"alloca~0.i", i32 3
  store i32 %"49~inst.i", ptr %"gep~57.i", align 4
  call void @llvm.lifetime.end.p0(i64 8, ptr %"alloca~0.i")
  call void @llvm.lifetime.end.p0(i64 4, ptr %"alloca~1.i")
  call void @llvm.lifetime.end.p0(i64 4, ptr %"alloca~2.i")
  store ptr %"36~inst", ptr %"alloca~0", align 8
  ret void
}

define void @"callable:[std/str.wo]str.constructor;methodType:classType:[std/str.wo]str<>#(usizeType,usizeType)=>voidType"(ptr %0, i32 %1, i32 %2) {
entry:
  %"alloca~0" = alloca ptr, align 8
  store ptr %0, ptr %"alloca~0", align 8
  %"alloca~1" = alloca i32, align 4
  store i32 %1, ptr %"alloca~1", align 4
  %"alloca~2" = alloca i32, align 4
  store i32 %2, ptr %"alloca~2", align 4
  %"gep~54" = getelementptr { i32, i32, i32, i32 }, ptr %"alloca~0", i32 2
  %"45~inst" = load i32, ptr %"gep~54", align 4
  %"46~inst" = load i32, ptr %"alloca~1", align 4
  %"gep~55" = getelementptr { i32, i32, i32, i32 }, ptr %"alloca~0", i32 2
  store i32 %"46~inst", ptr %"gep~55", align 4
  %"gep~56" = getelementptr { i32, i32, i32, i32 }, ptr %"alloca~0", i32 3
  %"48~inst" = load i32, ptr %"gep~56", align 4
  %"49~inst" = load i32, ptr %"alloca~2", align 4
  %"gep~57" = getelementptr { i32, i32, i32, i32 }, ptr %"alloca~0", i32 3
  store i32 %"49~inst", ptr %"gep~57", align 4
  ret void
}

declare noalias ptr @malloc(i32) #0

; Function Attrs: nocallback nofree nosync nounwind willreturn memory(argmem: readwrite)
declare void @llvm.lifetime.start.p0(i64 immarg, ptr nocapture) #1

; Function Attrs: nocallback nofree nosync nounwind willreturn memory(argmem: readwrite)
declare void @llvm.lifetime.end.p0(i64 immarg, ptr nocapture) #1

attributes #0 = { "alloc-family"="malloc" "allockind"="alloc,zeroed" "allocsize"="0" }
attributes #1 = { nocallback nofree nosync nounwind willreturn memory(argmem: readwrite) }
